### PR TITLE
[WIP] Search widgets

### PR DESCRIPTION
**Pull request by carolinux**
For better filter by column selection -> having the option to filter by dropdown menu etc when your data is such.

### CODE REVIEW

File: src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.cpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
+/***************************************************************************
+    qgstexteditwrapper.cpp
+     --------------------------------------
+    Date                 : 5.1.2014
+    Copyright            : (C) 2014 Matthias Kuhn
```
Comment (3nids) [inline src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.cpp]: You can write your name here ;)

File: src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+QgsDefaultSearchWidgetWrapper::QgsDefaultSearchWidgetWrapper( QgsVectorLayer* vl, int fieldIdx, QWidget* editor, QWidget* parent )
+    : QgsEditorWidgetWrapper( vl, fieldIdx, editor, parent )
+    , mLineEdit( NULL )
+{
+}
+
+QVariant QgsDefaultSearchWidgetWrapper::value()
+{
+    return QString("foo");
```
Comment (3nids) [inline src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.cpp]: really?

File: src/gui/editorwidgets/core/qgseditorwidgetregistry.cpp [hunk 3]
```diff
@@ -100,6 +100,25 @@ QgsEditorWidgetWrapper* QgsEditorWidgetRegistry::create( const QString& widgetId
   return 0;
 }
 
+QgsEditorWidgetWrapper* QgsEditorWidgetRegistry::createSearch( const QString& widgetId, QgsVectorLayer* vl, int fieldIdx, const QgsEditorWidgetConfig& config, QWidget* parent, const QgsAttributeEditorContext &context )
```
Comment (3nids) [inline src/gui/editorwidgets/core/qgseditorwidgetregistry.cpp]: wouldn't `createSearchWidget` be more appropriate and coherent?

**Comment by 3nids** [comment]
Good job!!!

There is a small UI problem: the line edit does not take the room in the attribute table.
image

Waiting for the value relation and relation reference widgets!

**Comment by carolinux** [comment]
Yeah, I see this UI thing too. Any ideas how to fix?

File: src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.cpp (hunk 2 above)
Comment (carolinux) [inline src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.cpp]: that was just for testing

**Comment by 3nids** [comment]
have a look at [QSizePolicy](http://doc.qt.io/qt-4.8/qsizepolicy.html)

**Comment by carolinux** [comment]
Couldn't get this to work. I think since here we have a placeholder for a widget which changes I can use stackedwidget. What do you think?
 [1]http://pyqt.sourceforge.net/Docs/PyQt4/qstackedwidget.html

File: src/app/qgsattributetabledialog.cpp [hunk 4]
```diff
@@ -407,7 +410,26 @@ void QgsAttributeTableDialog::filterColumnChanged( QObject* filterAction )
...
+  int fldIdx = mLayer->fieldNameIndex( fieldName );
+    if ( fldIdx < 0 )
+      return;
+  const QString widgetType = mLayer->editorWidgetV2( fldIdx );
+  const QgsEditorWidgetConfig widgetConfig = mLayer->editorWidgetV2Config( fldIdx );
+ //replace with createSearch or so
+ //go to registry and create a create Search method
+  mCurrentSearchWidgetWrapper= QgsEditorWidgetRegistry::instance()->createSearch( widgetType, mLayer, fldIdx, widgetConfig, mFilterContainer);
+  mCurrentSearchWidgetWrapper->widget()->setObjectName("searchy");
+  mCurrentSearchWidgetWrapper->widget()->setVisible( true );
```
Comment (3nids) [inline src/app/qgsattributetabledialog.cpp]: that would be the place to make it larger I think.
can you try `setSizePolicy(QSizePolicy::Expanding)` ?

Comment (carolinux) [inline src/app/qgsattributetabledialog.cpp]: I tried this, didn't work. I made a function to replace the widgets in the layout and this seems to work.

**Comment by carolinux** [comment]
Looks good for value maps too: http://imgur.com/SrIvKKu

**Comment by 3nids** [comment]
perfect!

**Comment by m-kuhn** [comment]
Works fine in some first testing, thank you very much @carolinux

File: src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+#include "qgseditorwidgetwrapper.h"
+#include <qgsfilterlineedit.h>
+
+
+/**
+ * Wraps a search widget. Default form is just a QgsLineFilterEdit 
+ *
+ */
+
+class GUI_EXPORT QgsDefaultSearchWidgetWrapper : public QgsEditorWidgetWrapper
```
Comment (m-kuhn) [inline src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.h]: I think this does not have to inherit from QgsEditorWidgetWrapper (the "Editor" part of the name indicates that it's there to edit fields which this widget actually does not) , I would inherit directly from QObject

File: src/gui/editorwidgets/core/qgseditorwidgetregistry.cpp [hunk 6]
```diff
@@ -16,7 +16,7 @@
 #include "qgseditorwidgetregistry.h"
 
 #include "qgsattributeeditorcontext.h"
-#include "qgseditorwidgetfactory.h"
+//#include "qgseditorwidgetfactory.h"
```
Comment (m-kuhn) [inline src/gui/editorwidgets/core/qgseditorwidgetregistry.cpp]: If it's not required, delete it :)

**Comment by m-kuhn** [comment]
The code looks good in general. Nice work!

Some hints:

* There is a script `./scripts/prepare-commit.sh` if you run it before committing it will perform some styling (whitespacing etc) so the code matches the rest of the QGIS codebase. I have it automatically performed before every commit by adding it to `.git/hooks/pre-commit`
* Instead of inheriting directly from QgsEditorWidgetWrapper it would be great to have a QgsSearchWidgetWrapper that offers a method `QString expression()` and a signal `expressionChanged(const QString&)` which give access to the filter.
* It would be nice if in the ValueMap search widget one would not have to click accept but only choose a column. Basically that would be to make use of the `expressionChanged()` signal mentioned above.
* If you are still looking for more use-cases, have a look at the ValueRelation and RelationReference widgets which also offer comboboxes, but based on features (rows) from other layers instead of a fixed map.

File: src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.h (hunk 5 above)
Comment (carolinux) [inline src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.h]: Or some other abstract base class like SearchWidgetWrapper?

Comment (m-kuhn) [inline src/gui/editorwidgets/core/qgsdefaultsearchwidgetwrapper.h]: Exactly, when I wrote this comment I had not completely understood your approach, so better see the comments in the PR.

**Comment by carolinux** [comment]
Alright, I will address the comments in a new PR in the weekend.

**Comment by m-kuhn** [comment]
Great! Looking forward.

**Comment by 3nids** [comment]
@carolinux it would be nice to fix this in the bugfix period. Do you have time to implement this?

**Comment by carolinux** [comment]
@3nids when is the end of the bugfix period? I only realistically have time in the weekends for the following 2 weeks

**Comment by 3nids** [comment]
the bugfix will end on the [26th](https://www.qgis.org/fr/site/getinvolved/development/index.html#release-schedule)
please let me know if you have time for this, or I can see with @m-kuhn to finish this.

**Comment by carolinux** [comment]
I definitely have time until then!

**Comment by 3nids** [comment]
perfect, thanks!

**Pull request merged**